src/execution/operator/persistent/physical_insert.cpp

```diff
@@ -309,24 +309,196 @@ static void RegisterUpdatedRows(InsertLocalState &lstate, const Vector &row_ids,
 	for (idx_t i = 0; i < count; i++) {
 		auto result = updated_rows.insert(data[i]);
 		if (result.second == false) {
+			// This is following postgres behavior:
 			throw InvalidInputException(
 			    "ON CONFLICT DO UPDATE can not update the same row twice in the same command. Ensure that no rows "
 			    "proposed for insertion within the same command have duplicate constrained values");
 		}
 	}
 }
 
+static void CheckDistinctnessInternal(DataChunk &input, const unordered_set<column_t> &column_ids,
+                                      map<idx_t, vector<idx_t>> &result) {
+	// FIXME: do we want to use the same validity mask across multiple distinctness checks (for different column_ids
+	// inputs) ?
+	ValidityMask valid(input.size());
+	for (idx_t i = 0; i < input.size(); i++) {
+		if (!valid.RowIsValid(i)) {
+			// Already a conflict
+			continue;
+		}
+
+		bool has_conflicts = false;
+		for (idx_t j = i + 1; j < input.size(); j++) {
+			if (!valid.RowIsValid(j)) {
+				// Already a conflict
+				continue;
+			}
+			bool matches = true;
+			for (auto &col_idx : column_ids) {
+				auto this_row = input.GetValue(col_idx, i);
```

Comment:
This code currently uses `Value` for distinctness - which is rather inefficient. Instead, I would suggest we can switch this to using sort keys, see e.g. `src/include/duckdb/core_functions/create_sort_key.hpp` and `CreateSortKeyHelpers`, we can then place those sort keys in a `string_map_t` and use that to find duplicates.

Reply:
Ah this was the first thing I changed, I didn't benchmark it, but a memcmp should be trivial in comparison to potentially nested function calls in the equality operator, especially given the amount of comparisons we perform here.

I didn't use a `string_map_t`, I prepare the sort_keys for each column separately as I could have multiple different combinations of `column_ids` in the non-"DO UPDATE" case, since we match every relevant index in that case